azure-devops/install_msvc_preview.ps1

```diff
@@ -0,0 +1,64 @@
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+$ErrorActionPreference = "Stop"
+
+Function InstallVS
+{
+    # Microsoft hosted agents do not have the required MSVC Preview for building MSVC STL.
+    # This step installs MSVC Preview, only on Microsoft hosted agents.
+
+  Param(
+    [String]$WorkLoads,
+    [String]$Sku,
+    [String]$VSBootstrapperURL)
+  $exitCode = -1
+  try
+  {
+    Write-Host "Downloading bootstrapper ..."
+    [string]$bootstrapperExe = Join-Path ${env:Temp} ([System.IO.Path]::GetRandomFileName() + ".exe")
+    Invoke-WebRequest -Uri $VSBootstrapperURL -OutFile $bootstrapperExe
+
+    $Arguments = ('/c', $bootstrapperExe, $WorkLoads, '--quiet', '--norestart', '--wait', '--nocache' )
+
+    Write-Host "Starting Install: $Arguments"
+    $process = Start-Process -FilePath cmd.exe -ArgumentList $Arguments -Wait -PassThru
+    $exitCode = $process.ExitCode
+
+    if ($exitCode -eq 0 -or $exitCode -eq 3010)
+    {
+      Write-Host -Object 'Installation successful!'
+    }
+    else
+    {
+      Write-Host -Object "Non zero exit code returned by the installation process : $exitCode."
+    }
+  }
+  catch
+  {
+    Write-Host -Object "Failed to install Visual Studio!"
+    Write-Host -Object $_.Exception.Message
+    exit $exitCode
+  }
+
+  exit $exitCode
+}
+
+# Invalidate the standard installation of VS on the hosted agent.
+Move-Item "C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/" "C:/Program Files (x86)/Microsoft Visual Studio/2019/nouse/" -Verbose
```

Comment:
This seems highly invasive. Is it desirable? Could we simply fail if we detect an existing installation of VS?

Reply:
This script is only run on the "hosted" machines which are blown away after each build, so invasive changes are fine/normal.

Reply:
then you would always fail on the hosted one.